### ISSUE TITLE
Error in EATA implementation

### ISSUE THREAD

**Issue by blaze1522** [oripress/CCC#5]
Dear authors,

When I ran EATA, it was running for a long time without giving any feedback. I found that the generation of Fisher weights was taking a long time. Upon further investigation, I found that the loader being used for generating fisher weights was the CCC loader. 

```python
  def get_fisher_vectors(self, model, fisher_amount, loader):
      model = configure_model(model)
      params, param_names = collect_params(model)
      ewc_optimizer = torch.optim.SGD(params, 0.001)
      fishers = {}
      train_loss_fn = nn.CrossEntropyLoss().cuda()
      for iter_, (images, targets) in enumerate(loader, start=1):
          images = images.cuda(non_blocking=True)
          outputs = model(images)
          ...  
```

I don't think this is correct behaviour, as the source/in-distribution test loader should be used to generate fisher weights.

Is the above a correct observation or am I doing something wrong? Kindly help.

Thanks

**Comment by oripress** [comment]
Hi,
You're right that our implementation differs from the EATA implementation. 
We explain why this is in Appendix C in the [paper.](https://arxiv.org/pdf/2306.05401.pdf)
The short story is that using source images is explicitly not allowed in test time adaptation. 

Hope this helps.

**Comment by blaze1522** [comment]
Got it. It makes sense. I think there is some indentation issue around here, because of which the loop is not breaking.

```python
def get_fisher_vectors(self, model, fisher_amount, loader):
                    ...
                    if iter_ * 64 >= fisher_amount:
                        fisher = fisher / iter_
                    fishers.update({name: [fisher, param.data.clone().detach()]})
                    break
                    ...
```

**Comment by oripress** [comment]
You're right, this is fixed now, thanks for pointing this out :)

**Comment by blaze1522** [comment]
Hi,

Sorry for troubling again. The previous code was actually almost correct. Just that the break needs to be one indentation outside and inside and if condition. The code below is working fine for me. The correct eata.py is bugged because ` fishers.update({name: [fisher, param.data.clone().detach()]})` is never triggered. We need the final explicit break because in general the fisher loader ends in the normal implementation but here it doesn't.

```python 
    def get_fisher_vectors(self, model, fisher_amount, loader):
        model = configure_model(model)
        params, param_names = collect_params(model)
        ewc_optimizer = torch.optim.SGD(params, 0.001)
        fishers = {}
        train_loss_fn = nn.CrossEntropyLoss().cuda()
        for iter_, (images, targets) in enumerate(loader, start=1):
            print(iter_)
            images = images.cuda(non_blocking=True)
            outputs = model(images)
            _, targets = outputs.max(1)
            loss = train_loss_fn(outputs, targets)
            loss.backward()
            for name, param in model.named_parameters():
                if param.grad is not None:
                    if iter_ > 1:
                        fisher = (
                            param.grad.data.clone().detach() ** 2 + fishers[name][0]
                        )
                    else:
                        fisher = param.grad.data.clone().detach() ** 2
                    if iter_ * 64 >= fisher_amount: # loader uses a batch size of 64
                        fisher = fisher / iter_
                    fishers.update({name: [fisher, param.data.clone().detach()]})
            ewc_optimizer.zero_grad()
            if iter_ * 64 >= fisher_amount:
                break
        del ewc_optimizer
        del loader

        return fishers
```

**Comment by oripress** [comment]
Thanks for this, I updated the EATA code.
